### ISSUE TITLE
Integrating alice/stages

### ISSUE THREAD

**Issue by lillian-lemmer** [hypatia-engine/hypatia#86]

**Comment by ejmr** [comment]
I'll go through it all over the weekend.

**Comment by lillian-lemmer** [comment]
I talked to my friend xax about `alice/stages` a bit. Below is a verbatim copy of what they said.

well i mean from hearing you describe it... like the kind of 'state' i think of when i think of 'in a menu vs. in a map' and the kind of 'state' i think of when i think of 'in a map vs. in a different part of the map' seem pretty distinct

iirc i did a similar thing with phantomnation way back in the day and i think what it ended up being was that... fundamental game states were useful for when you needed different input mapping (e.g., esc to cancel out of menus while in a menu vs. esc to open the system menu when in a map) but the rest of the time it was better to handle actual game divisions using some other mechanism

like e.g., having a map trigger to switch between "inside" and "outside" views, or to load the next level or w/e

rly my Big Advice is: never code something into the engine that you don't immediately use in two+ places in the actual game

b/c it's very easy to think up Engine Features that seem cool or useful only to realize after they're there that they have weird implications or require weird assumptions or just... don't work well

b/c that is the problem i make incessantly myself really :V

**Comment by ejmr** [comment]
> rly my Big Advice is: never code something into the engine that you don't immediately use in two+ places in the actual game

I strongly support and share this sentiment.  From my own experience working on game engines, *do not* spend time trying to guess what functionality *might* be useful in the future.  That very easily becomes a time-sink and, even worse, wasted effort if the functionality goes unused.

**Comment by ejmr** [comment]
In my opinion the branch needs a decent amount of reworking before merging.  It has some great ideas and functionality which is unfortunately masked by problems with overall structure and organization.  The rest of this comment is a copy of the notes I wrote while reviewing the branch.

# Notes on `alice/stages`

Written by Eric James Michael Ritz on 30 October 2015.  This document organizes comments by the commit which includes them.

## 4fa4a22c28be37b978a9560378804cd369db4be9

The commit message provides no description of "stages" or a justification for the patch in general.

### `hypatia/constants.py`

This file doesn't "feel" like the appropriate place for `DEFAULT_CONFIG`.  Configuration files are so common in game engines that it is reasonable to expect Hypatia to have a `hypatia/config.py` file, or something along those lines.  Moving `DEFAULT_CONFIG` into such a file would be better organization.  The `constants.py` file is also for immutable data.  While the default configuration itself could be immutable, that does not alter the fact that the engine is free to change configuration values elsewhere, e.g. when loading the preferred configuration of a player.  This makes `DEFAULT_CONFIG` "not constant enough" to warrant being a part of `constants.py`.

### `hypatia/game.py`

The `Stage` class has no docstring, which is mandatory per [the project guidelines](http://hypatia-engine.github.io/wiki/contributing.html#docstrings).

Various docstrings throughout the file seem to use the terms "stage" and "scene" interchangably.  It would be better to pick one term and stick with it consistently.  However, if "stages" and "scenes" are not synonymous then the code needs to document the differences.

The `ExceptionStage` class, simply from its name, sounds like a class meant for use with `raise`; it needs a docstring explaining this.  The implementation makes it seem that the purpose of the class is to be a "stage" meant for displaying exception information within a game while it runs.  So for that reason a name like `ExceptionDisplayingStage` may be better to improve readability and help avoid confusion with regard to incorrectly thinking of `ExceptionStage` as a type of an exception.

Finally, stages feel complex enough to warrant their own module, e.g. `hypatia/stages.py`.

## 1b4dcf6a72c8c5ea409cdadcf1b0d5798e86b0da

The commit changes every line in two files, making it almost impossible to see the actual 2--4 which constitute the real changes of this patch.  The commit needs to be rewritten so that it only changes the lines that are actually relevant to the patch.

## 684cbacd90c3bf67541e8a21377ad8128b8434fe

No classes in `hypatia/vfs.py` have docstrings.

The exceptions should have `Exception` in their name, e.g. `IsADirectory` should be `IsADirectoryException`, for consistency with all other custom exceptions in the engine.

Overall though the implementation in `hypatia/vfs.py` is well-written.  All it needs is docstrings and ideally some docstring unit tests.  But the code itself is well-structured and easy to read and understand.

## fe302fe1af68432ae63907f2d9ed9e40ebcb5a48

Since this commit fixes bugs from the previous commit it should simply be squashed into `684cbacd` via an interactive rebase in Git.

## d185f1163688cb43acbf76e5758dd6041751fdeb

### `hypatia/game.py`

Regarding this line:

```python
self.config.readfp(self.vfs.open("/game/game.ini"))
```

Is there a potential chance that the engine will interpret `/game/game.ini` as an absolute filepath?  If the path should always be relative then `./game/game.ini` would make that clear to developers.

The comments for the `setup()` method would be better as a docstring, as in general any comments immediately following a method definition *usually* belong in a docstring.

### `hypatia/scene.py`

The file contains no documentation whatsoever, which exacerbates the confusion between "stage" and "scene".

## 93c1fbf5c7375c9d0d4f075e873e4ec4dd6ecc29

This is another commit which must be squashed before merging the branch.  As a side note, when developers find a bug in their most recent commit they can add the bug-fixes directly to the last commit by using `git add -u; git commit --amend`.

## 4043781d5e2eb6c6786d8a58a865dc1ec7731c1d

The commit messages asks how the `--user` flag made it into the `script/bootstrap` file in the first place.  The commit message should explain that by using `git log -p -S"--user" --all -- script/bootstrap`, aka. using "the pickaxe" in Git terms.  That command shows that commit `4fa4a22c` introduced the `--user` flags.

## 939e8c074a757de563a12ecbd9388b5e8c15288e

The commit message should explain "exception scrolling" and why it belongs in the engine.  It should not have a message asking for help with testing, which is more appropriate for the relevant GitHub issue and/or pull-request page for the branch.

## 0541d17eb9a9c78c30bf9f26e6bc608eb21df074

The creation of `hypatia/util.py` is a great idea.

**Comment by lillian-lemmer** [comment]
@alxcenz have you read through these notes?

**Comment by alxcenz** [comment]
Yeah, I've been working through this. Expect to see me rebase this on top of latest develop and squish a few remaining bugs, as well as implementing the game logic in a new state soon.
